api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,203 @@
+syntax = "proto3";
+
+package envoy.service.mutation.v3alpha;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/service/auth/v3/attribute_context.proto";
+import "envoy/type/v3/http_status.proto";
+
+import "google/protobuf/struct.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.mutation.v3alpha";
+option java_outer_classname = "MutationProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: External Mutation Service]
+
+// A service that can mutate HTTP requests and responses.
+// The overall mutation protocol works like this:
+//
+// 1. Envoy sends to the service information about the HTTP request.
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server a chunk of the message body.
+//    and the server sends back a MutationResponse.
+// 4. If so requested, Envoy sends the server the HTTP trailers,
+//    and the server sends back a MutationResponse.
+// 5. At this point, request processing is done, and we pick up again
+//    at step 1 when Envoy receives a response from the upstream server.
+// 6. At any point above, if the server closes the gRPC stream cleanly,
+//    then Envoy proceeds without consulting the server.
+// 7. At any point above, if the server closes the gRPC stream with an error,
+//    then Envoy returns a 500 error to the client, unless the filter
+//    was configured to ignore errors.
+//
+// In other words, the process is a request/response conversation, but
+// using a gRPC stream to make it easier for the client and server to
+// maintain state. The exception is the "SEND_BODY_ASYNC" command,
+// which requests that the server keep sending body chunks without waiting
+// for a response.
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+message MutationRequest {
+  oneof r {
+    // Information about the HTTP request, sent after the HTTP headers have been
+    // received on the "encoding" path, but before the body is available.
+    // This is the same data sent to an "ext_authz" service.
+    auth.v3.AttributeContext request = 1;
```

Comment:
Not sure I fully follow why this is only available when the upstream is repsonding and we're encoding.. why can't we have request metadata or attributes earlier?

Reply:
I was just trying to explain that this message is sent when decodeHeaders is called on the filter. I'll see about changing the comment.

Reply:
I think in general referencing Envoy filter lifecycle events makes sense in this file, rather than more abstract concepts of request/response, since for streaming some of these break down.